File: credit.c
//Libraries
#include <stdio.h>
#include <cs50.h>

//Delcaring functions
int get_card_length(long);
string get_card_type(long);
bool verify_card_checksum(long);

//Main function 
int main(void)
{
    long card_number = get_long("Number: ");
    int card_length = 0;
    string card_type = "INVALID";
    bool checksum_valid = false;
    card_type = get_card_type(card_number);
    if (verify_card_checksum(card_number) == false) //only verify checksum if provider is valid
    {
        card_type = "INVALID";
    }
    printf("%s\n", card_type);
}


//Get the number of digits in the card
int get_card_length(long card_number)
{   
    int i = 0;
    while (card_number >= 1)
    {
        card_number = card_number / 10;
        i++;
    }
    return i;
}


//Get the provider of the credit card
string get_card_type(long card_number)
{   
    //Get first two digits
    string a = "INVALID";
    int b = get_card_length(card_number);
    int first_two_digits;
    for (int x = get_card_length(card_number) - 2; x > 0; x--)
    {
        card_number = card_number / 10;
    }
    first_two_digits = card_number;
    //Determine card provider to digit lengths
    if (first_two_digits == 34 || first_two_digits == 37)
    {
        if (b == 15)
        {
            a = "AMEX";
        }
    }
    else if (first_two_digits > 50 && first_two_digits < 56)
    {
        if (b == 16)
        {
            a = "MASTERCARD";
        }
    }
    else if (first_two_digits > 39 && first_two_digits < 50)
    {
        if (b == 13 || b == 16)
        {
            a = "VISA";
        }
    }
    else
    {
        a = "INVALID";
    }
    return a;
}


//Verify that the card checksum is valid
bool verify_card_checksum(long card_number)
{
    int check_sum = 0;
    //Get the checksum
    while (card_number >  0)
    {
        //odd digits
        check_sum += card_number % 10;
        card_number /= 10;
        //even digits
        if ((card_number % 10) < 5)
        {
            check_sum += (card_number % 10) * 2;
        }
        else
        {
            check_sum += (((card_number % 10) * 2) - 10) + 1;
        } 
        card_number /= 10;
    }
    //Verify the checksum
    if (check_sum % 10 == 0)
    {
        return true;        
    }
    else
    {
        return false;        
    }
}

File: pset6/credit.py
import re
from cs50 import get_int


def main():
    # get card number from user
    card_number = get_int("Number: ")
    # get the provider of the card based on numbers
    card_provider = provider(card_number)
    # if card number is valid, check the number
    if card_provider != "INVALID":
        if verify_number(card_number) is False:
            card_provider = "INVALID"
    print(card_provider)

# takes the card number and if it matches the starting numbers and number length returns the provider, otherwise invalid


def provider(card_number):
    if re.search("^34|^37", str(card_number)):
        if len(str(card_number)) == 15:
            return "AMEX"
    elif re.search("^5[1-5]", str(card_number)):
        if len(str(card_number)) == 16:
            return "MASTERCARD"
    elif re.search("^4", str(card_number)):
        if len(str(card_number)) == 13 or len(str(card_number)) == 16:
            return "VISA"
    return "INVALID"


def verify_number(card_number):
    # initialize checksum to 0
    checksum = 0
    # for the digits of the card, alternate between adding the last digit to checksum, and 2x the last digit
    # start at 0, end at length of the card, step by 2 since you check 2 digits at a time
    for i in range(0, len(str(card_number)), 2):
        # //odd digits
        checksum += card_number % 10
        # // makes it do integer division
        card_number //= 10
        # even digits, if its greater then 10, add one and subtract 10 to get the sum of the digits (14 = 1 + 4 = 5)
        if ((card_number % 10) < 5):
            checksum += (card_number % 10) * 2
        else:
            checksum += (((card_number % 10) * 2) - 9)
        card_number //= 10
    # if checksum passes return true, else false
    if checksum % 10 == 0:
        return True
    return False


main()
